### DISCUSSION TITLE
The results after LoRA training are very disappointing...

### DISCUSSION THREAD

**Discussion by ArachneRomBor** [oobabooga/text-generation-webui#2353] [Q&A]
Hello.

Like many others, I'm also trying to delve into the secrets of GPT and for my first attempts I choose this project `oobabooga/text-generation-webui`.
Unfortunately, the results after LoRA training are very disappointing...

At the outset, I have to say that I don't have a suitable graphics card yet, so for the first attempts I had to choose an installation (for Windows) with only CPU support.

I choose `huggyllama/llama-7b` as a model and then added `oobabooga/llama-tokenizer`.

After starting the server
`python server.py --load-in-8bit --cpu`
I selected `huggyllama/llama-7b` in the `Model` tab and it loaded successfully.
![webui-model-loaded](https://github.com/oobabooga/text-generation-webui/assets/134591923/6c145ec9-c20c-4437-b12c-77b31edeaf7a)

On the `Training` tab, I set the name of the LoRA file, I set the LoRA rank to 128 and the LoRA Alpha to 256 (they should have, according to the instructions
(https://github.com/oobabooga/text-generation-webui/blob/main/docs/Training-LoRAs.md),
for raw text higher values than default and Alpha should be twice the Rank).
Subsequently, I selected the prepared training file on the `Raw text file` sub-tab and started LoRA training.

After the training, I restarted the server in `chat` mode
`python server.py --load-in-8bit --cpu --chat`,
I loaded the model again and at the same time selected the generated LoRA from the previous training and connected to the model using `Apply the selected LoRAs`.
![webui-model-lora-loaded](https://github.com/oobabooga/text-generation-webui/assets/134591923/b52479d4-7a10-4214-b0ba-0ee95c56e128)

After a long description comes the essentials.
I expected (apparently naively) that the assistant would at least partially know the newly learned information, but he was literally making things up.
I tried repeating the training with the original text and keeping the Rank and Alpha values as recommended in the mentioned tutorial, changing only the Epochs and Learning Rate, but the result was very similar, although the Loss value always decreased and was less than 1.

Am I doing anything wrong?
Was my procedure correct?
Or I didn't even understand how the whole thing is supposed to work?
Could the unsatisfactory results be related to the fact that I'm only running in CPU mode?

Thanks in advance for any ideas to move me forward…

**Comment by hiyata** [comment]
Having the same issues training on 3090 in 8bit following the instructions posted. I'm not sure if it's an issue with the actual text generation, the prompt settings, or with the training settings. I've wondered if the LoRa is even loading or I'm just that bad at figuring out prompting. Have you been able to fix it?

**Comment by caterpillarpants** [comment]
Answer1:
If your bot is learning, but not well, you can increase lora rank, epochs or learning rate. Sounds like you are already hip to this.

If you increase one of these parameters and the bot forgets other stuff it used to know, the parameter has increased too much. Since this does not appear to be the case, you are go to increase further. sounds like the bot knows the material, but not well. I would continue increasing epochs and lower learning rate. which should give it finer knowledge of the subject. 

And increasing learning rate is a good thing to try since it doesn't increase ram usage. This makes it more likely that it will remember something it reads. If your bot starts making wild assumptions, this may be too high.

If the thing is not getting that it should be accessing the info you trained it on when prompted, perhaps your lora rank is too low, but it sounds like it knows stuff, but is just confused so I wouldn't suspect lora alpha in this case.

Good luck, it may take some tweaking to get a quality train, but in my experience it definitely learns and it takes time to get a feel for the parameters. Have fun!


Answer2: change the input hyperparameters to something precise, if you haven't already.


Disclaimer: I am a hobbyist and likely have no idea what I'm talking about, but this is my experience with oobabooga.

**Reply by tmm1** [reply]
thanks for sharing your experience! there's not a lot of information available about fine-tuning with lora.

can you share any examples of what you've tried and what kind of results you've had?

and what do you mean by "change the input hyperparameters" where would that happen

**Comment by vdruts** [comment]
Have you guys figured this out? 

Curious about how would one go about training a 'style' lora... What type of things would the data require, how would question/answer work with writing style.

**Reply by caterpillarpants** [reply]
use a low rank like 8. instead of the dataset, just feed it raw text to see if that works. If you insist on a q&a dataset, don't worry if you are just going for style. Just make all the answers in the style you want.
